pkg/csi/service/wcp/controller.go

```diff
@@ -954,7 +958,69 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 			return nil, csifault.CSIInvalidArgumentFault, err
 		}
 		volumeType = prometheus.PrometheusBlockVolumeType
-
+		volumeAttachment, err := commonco.ContainerOrchestratorUtility.GetVolumeAttachment(ctx, req.VolumeId, req.NodeId)
+		var isVADeleted bool
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Info("VolumeAttachments object not found for volume %q & node %q. "+
+					"Thus, assuming the volume is detached.", req.VolumeId, req.NodeId)
+				isVADeleted = true
+			} else {
+				log.Errorf("failed to retrieve volumeAttachment from API server Err: %v:", err)
+				return nil, csifault.CSIInternalFault, err
+			}
+		}
+		if !isVADeleted {
+			log.Infof("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
+			for k, v := range volumeAttachment.Status.AttachmentMetadata {
+				if k == common.AttributeVmUuid {
+					log.Debugf("controllerUnpublishVolumeInternal: vmuuid value: %q", v)
+					vcdcMap, err := getDatacenterFromConfig(c.manager.CnsConfig)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to get datacenter from config with error: %+v", err)
+					}
+					var vCenterHost, dcMorefValue string
+					for key, value := range vcdcMap {
+						vCenterHost = key
+						dcMorefValue = value
+					}
+					vc, err := c.manager.VcenterManager.GetVirtualCenter(ctx, vCenterHost)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"cannot get virtual center %s from virtualcentermanager while attaching disk with error %+v",
+							vc.Config.Host, err)
+					}
+					podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to the PodVM Moref from the PodVM UUID: %s in datacenter: %s with err: %+v",
+							v, dcMorefValue, err)
+					}
+					isStillAttached := false
+					timeout := 4 * time.Minute
+					pollTime := time.Duration(5) * time.Second
+					err = wait.Poll(pollTime, timeout, func() (bool, error) {
+						diskUUID, err := cnsvolume.IsDiskAttached(ctx, podVM, req.VolumeId, true)
+						if err != nil {
+							return false, err
+						}
+						if diskUUID != "" {
+							log.Errorf("diskUUID: %q is still attached", diskUUID)
```

Comment:
May be enhance the log to indicate that it's being retried. And will an info log make sense?

Reply:
Addressed it in the new PR #1702